### PR TITLE
Centralized training of tabular data

### PR DESCRIPTION
**Pull request by sanaAyrml**
# PR Type
[Other]

# Short Description

Add a small reference implementation for centralized training of tabular data

### CODE REVIEW

File: vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb
[the remark this answers is not in the document]
Comment (fatemetkl) [reply vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: I ran the code, and it works alright, but I faced an error in one line and had to change something slightly. It might just be on my side, though. 
The issue was that `pd.get_dummies()` by default returns `numpy.bool_` (and it seems to be a recent change), which seems like it is not a type accepted to be converted to a torch tensor. I am attaching the error screenshot here. What fixed it for me was changing the default type to int:`df = pd.get_dummies(df, columns=cols, dtype=int)`
[source](https://github.com/mrdbourke/tensorflow-deep-learning/issues/559)
Screenshot 2025-03-26 at 3 12 43 PM

File: vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb [hunk 1]
```diff
@@ -0,0 +1,952 @@
+{
```
Comment (nerdai) [inline vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>Maybe beneath this cell or above it, we can add a "Open in Google Colab" badge. To generate one, just use this: <a href="https://openincolab.com" target="_blank">https://openincolab.com/</a></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (nerdai) [inline vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>Does this notebook require the installation of <code>fl4health</code> library? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (nerdai) [inline vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #19.    
</pre><p>nit: do we need the 2 spaces here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (nerdai) [inline vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #31.    
</pre><p>nit: do we need the two spaces here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: pyproject.toml [hunk 2]
```diff
@@ -106,6 +106,25 @@ sphinx-design = "^0.6.1"
...
+types-colorama = "^0.4.15.20240311"
+types-defusedxml = "^0.7.0.20240218"
+types-gevent = "^24.11.0.20250305"
+types-mock = "^5.2.0.20250306"
+types-pexpect = "^4.9.0.20241208"
+types-psutil = "^7.0.0.20250218"
+types-python-dateutil = "^2.9.0.20241206"
+types-regex = "^2024.11.6.20250318"
+types-tensorflow = "^2.18.0.20250319"
+types-tqdm = "^4.67.0.20250319"
```
Comment (nerdai) [inline pyproject.toml:126]: Does this mean that in order to run the notebook, the user should install this group?

**Review by nerdai — COMMENTED** [review]
Looks in good shape! One thing I am wondering about is how we expect the user to set up their environment to run the notebook?

File: vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb
[the remark this answers is not in the document]
Comment (fatemetkl) [reply vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: Another minor comment: there seem to be some extra lines between some of the comments and their code. For example, `# Perform min-max scaling` and `def scale_min_max`.

**Review by fatemetkl — COMMENTED** [review]
Looks good overall! Just some comments.

File: pyproject.toml (hunk 2 above)
Comment (fatemetkl) [reply pyproject.toml:126]: I somehow ran the code without updating my venv :scream_cat:

**Comment by nerdai** [comment]
BTW, I added ReviewNB app quickly to review the notebook in this PR. We can remove after the fact if it becomes bothersome, or keep it if we anticipate having more notebooks checked into this repo.

File: pyproject.toml (hunk 2 above)
Comment (emersodb) [reply pyproject.toml:126]: This is likely due to the `--ignore-missing-imports` args that I added to the nbqa-mypy arg in our pre-commits. My guess is that @sanaAyrml likely added these types to avoid mypy throwing a fit about external types that we don't care about. I tried to get the nbqa-mypy to recognize our mypy.ini using the --config-file argument. However, it's still a pain in the butt about torch types for some reason, even though our standard mypy is fine. So I just decided we don't really care to much about mypy on notebooks anyway, since we've not even had a single notebook up to this point.

File: vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb (hunk 1 above)
Comment (sanaAyrml) [reply vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>No, this is more to train a centralized model to later bring to fl4health library.</p>

File: pyproject.toml (hunk 2 above)
Comment (sanaAyrml) [reply pyproject.toml]: so simple that we don't need much dependencies

File: vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb
[the remark this answers is not in the document]
Comment (sanaAyrml) [reply vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb:1]: I don't think adding a dataset description is essential but good idea to add a link to it. Thanks for mentioning.

File: vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb (hunk 1 above)
Comment (sanaAyrml) [reply vector-bootcamp-2025/tabular_ centralized_training/centralized_training_for_tabular_data.ipynb]: Added this

**Review by emersodb — COMMENTED** [review]
Looks almost ready. Just a few small comments.

File: vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb [hunk 3]
```diff
@@ -0,0 +1,952 @@
+{
```
Comment (emersodb) [inline vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>Perhaps its worth mentioning that, for their FL prototyping experiments, it will likely be best to preprocess the data before splitting it up across clients. However, in real world settings, it is also good to think about the challenges of feature preprocessing when the data is already separate.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (emersodb) [inline vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #26.    df = scale_min_max(df, "Age", train_age_min, train_age_max)
</pre><p>Maybe a silly question, but any reason we're not fitting a min-max scalar here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (emersodb) [inline vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>Super minor, but perhaps</p><p>Perceptrons -&gt; Perceptron</p><p>This model can be replaced with more complex and advanced models can be found in:" -&gt; &nbsp;This model can be replaced with more complex and advanced ones, which can be found in:</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (emersodb) [inline vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #32.            acc = accuracy_fn(y_train.int(), y_pred)
</pre><p>Is this int conversion necessary? It might be, but for some reason feels like it could be dropped to me.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (emersodb) [inline vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>We're saying here that we've going to evaluate the model on the test data, but we stop just short of that. That is, we gather all the predictions, but don't measure it. I think that's okay, but maybe we change the comment here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/VectorInstitute/FL4Health/pull/362/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

File: pyproject.toml [hunk 4]
```diff
@@ -121,6 +121,26 @@ sphinx-design = "^0.6.1"
 sphinxcontrib-apidoc = "^0.5.0"
 sphinx-autobuild = "^2024.10.3"
 
+
+[tool.poetry.group.bootcamp.dependencies]
+dm-tree = "^0.1.9"
+cmake = "^3.31.6"
+matplotlib = "^3.10.1"
+pandas-stubs = "^2.2.3.250308"
```
[the remark this answers is not in the document]
Comment (sanaAyrml) [reply pyproject.toml:129]: Sure let me drop it.

File: vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb (hunk 3 above)
Comment (sanaAyrml) [reply vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>Good Idea.</p>

Comment (sanaAyrml) [reply vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>I think it is a matter of choice. I believe it would have worked with standard scalar as well.</p>

Comment (emersodb) [reply vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: Sounds good. I think this works. Was just curious.

Comment (sanaAyrml) [reply vector-bootcamp-2025/tabular_centralized_training/centralized_training_for_tabular_data.ipynb:1]: <p>Currect, I think it is more for when labels are boolean. </p>

**Review by emersodb — APPROVED** [review]
Looks good to me.
